code/modules/detectivework/evidence.dm

```diff
@@ -1,94 +1,72 @@
-		user << "<span class='notice'>[I] won't fit in [src].</span>"
-		return
-
-	if(contents.len)
-		user << "<span class='notice'>[src] already has something inside it.</span>"
-		return
-
-	if(!isturf(I.loc)) //If it isn't on the floor. Do some checks to see if it's in our hands or a box. Otherwise give up.
-		if(istype(I.loc,/obj/item/weapon/storage))	//in a container.
-			var/obj/item/weapon/storage/U = I.loc
-			U.remove_from_storage(I, src)
-		if(user.is_holding(I))
-			user.dropItemToGround(I)
-		else
-			return
 
-	user.visible_message("[user] puts [I] into [src].", "<span class='notice'>You put [I] inside [src].</span>",\
-	"<span class='italics'>You hear a rustle as someone puts something into a plastic bag.</span>")
-
-	icon_state = "evidence"
-
-	var/xx = I.pixel_x	//save the offset of the item
-	var/yy = I.pixel_y
-	I.pixel_x = 0		//then remove it so it'll stay within the evidence bag
-	I.pixel_y = 0
-	var/image/img = image("icon"=I, "layer"=FLOAT_LAYER)	//take a snapshot. (necessary to stop the underlays appearing under our inventory-HUD slots ~Carn
-	img.plane = FLOAT_PLANE
-	I.pixel_x = xx		//and then return it
-	I.pixel_y = yy
-	add_overlay(img)
-	add_overlay("evidence")	//should look nicer for transparent stuff. not really that important, but hey.
-
-	desc = "An evidence bag containing [I]. [I.desc]"
-	I.loc = src
-	w_class = I.w_class
-	return 1
-
-/obj/item/weapon/evidencebag/attack_self(mob/user)
-	if(contents.len)
-		var/obj/item/I = contents[1]
-		user.visible_message("[user] takes [I] out of [src].", "<span class='notice'>You take [I] out of [src].</span>",\
-		"<span class='italics'>You hear someone rustle around in a plastic bag, and remove something.</span>")
-		cut_overlays()	//remove the overlays
-		user.put_in_hands(I)
-		w_class = WEIGHT_CLASS_TINY
-		icon_state = "evidenceobj"
-		desc = "An empty evidence bag."
-
-	else
-		user << "[src] is empty."
-		icon_state = "evidenceobj"
-	return
+/obj/item/weapon/storage/evidencebag/Initialize()
+	..()
+	RecalculateWeightClass(null)
+
+/obj/item/weapon/storage/evidencebag/can_be_inserted(obj/item/W, stop_messages = 0, mob/user)
+	. = ..()
+	if(.)
+		if(ismob(loc))
+			var/mob/M = loc
+			if(!M.get_held_index_of_item(src))
+				//can only insert if holding
+				user << "<span class='warning'>You must be holding \the [src] before adding to it!</span>"
+				return FALSE
+		else if(istype(loc, /obj/item/weapon/storage))
+			user << "<span class='warning'>Take \the [src] out of [loc] before adding to it!</span>"
+			return FALSE
+		return TRUE
+
+/obj/item/weapon/storage/evidencebag/handle_item_insertion(obj/item/W, prevent_warning = 0, mob/user)
+	. = ..()
+	RecalculateWeightClass(user)
+
+/obj/item/weapon/storage/evidencebag/proc/RecalculateWeightClass(mob/user)
+	var/old_wc = w_class
+	w_class = WEIGHT_CLASS_TINY
+	for(var/A in contents)
+		var/obj/item/I = A
+		if(istype(I))
+			w_class = max(w_class, I.w_class + 1)
+	if(user)
+		var/diff = abs(old_wc - w_class)

```

Comment:
Instead of all this abs() stuff, why not a switch() statement with -3,-2,-1,1,2,3